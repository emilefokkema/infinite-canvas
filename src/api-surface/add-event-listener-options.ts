/**
 * Options that can be provided when adding an event listener to an {@link InfiniteCanvas}
 */
export interface AddEventListenerOptions{
    /**
     * If `true`, the listener will be invoked at most once
     */
    once: boolean;
}