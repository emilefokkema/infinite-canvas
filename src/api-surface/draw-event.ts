import { Transformed } from "./transformed";

export interface DrawEvent extends Transformed{
}
