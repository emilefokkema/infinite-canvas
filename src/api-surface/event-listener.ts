import { EventMap } from "./event-map";


export declare type EventListener<K extends keyof EventMap> = (ev: EventMap[K]) => void;
