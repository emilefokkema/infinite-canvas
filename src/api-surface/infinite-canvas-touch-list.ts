import { InfiniteCanvasTouch } from "./infinite-canvas-touch";

export interface InfiniteCanvasTouchList extends TouchList{
    item(index: number): InfiniteCanvasTouch | null;
    [index: number]: InfiniteCanvasTouch;
}

export * from './infinite-canvas-touch';