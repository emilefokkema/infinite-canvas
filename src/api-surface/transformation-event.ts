import { Transformed } from "./transformed";

export interface TransformationEvent extends Event, Transformed{
}
