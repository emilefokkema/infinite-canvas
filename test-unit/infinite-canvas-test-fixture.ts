import { Mock, MockInstance, vi } from 'vitest'
import { DrawingLock } from "src/drawing-lock";
import { MockCanvasMeasurementProvider } from './mock-canvas-measurement-provider';
import { CanvasContextMock } from './canvas-context-mock';
import { Config } from 'api/config';
import { MockDrawingIterationProvider } from './mock-drawing-iteration-provider';
import { CssLengthConverterFactory } from 'src/css-length-converter-factory';
import { InfiniteCanvasViewBox } from 'src/infinite-canvas-viewbox';
import { RectangleManagerImpl } from 'src/rectangle/rectangle-manager-impl';
import { InfiniteContext } from 'src/infinite-context/infinite-context';
import { ViewBox } from 'src/interfaces/viewbox';
import { InfiniteCanvasRenderingContext2D } from 'api/infinite-canvas-rendering-context-2d';


function setupGlobals(){
	window.createImageBitmap = function(): Promise<ImageBitmap>{return undefined;};
	(<any>window).ImageData = class {
		public width: number;
		public height: number;
		public data: Uint8ClampedArray;
		constructor(arrayOrWidth: Uint8ClampedArray | number, widthOrHeight: number, height?: number){
			if(typeof arrayOrWidth === "number"){
				this.width = arrayOrWidth;
				this.height = widthOrHeight;
			}else{
				this.width = widthOrHeight;
				this.height = height;
				this.data = arrayOrWidth;
			}
		}
	};
}

export interface InfiniteCanvasTestFixture{
    width: number
    height: number
    releaseDrawingLockSpy: MockInstance<()=>void>
    getDrawingLockSpy: MockInstance<()=>DrawingLock>
    measurementProvider: MockCanvasMeasurementProvider
    contextMock: CanvasContextMock
    config: Partial<Config>
    mockDrawingIterationProvider: MockDrawingIterationProvider
    viewbox: ViewBox,
    infiniteContext: InfiniteCanvasRenderingContext2D
}

export function createInfiniteCanvasTestFixture(): InfiniteCanvasTestFixture{
    setupGlobals();
    const drawingLock: DrawingLock = {release(){}};
    const releaseDrawingLockSpy = vi.spyOn(drawingLock, 'release');
    const getDrawingLockSpy = vi.fn<()=>DrawingLock>().mockReturnValue(drawingLock)
    const width = 200;
    const height = 200;
    const measurementProvider = new MockCanvasMeasurementProvider(width, height);
    const contextMock = new CanvasContextMock();
    const config = {}
    const mockDrawingIterationProvider = new MockDrawingIterationProvider()
    const cssLengthConverterFactory: CssLengthConverterFactory = {
        create: () => ({
            getNumberOfPixels(value: number){
                return value;
            }
        })
    };
    const viewbox = new InfiniteCanvasViewBox(
        new RectangleManagerImpl(measurementProvider, config),
        contextMock.mock,
        mockDrawingIterationProvider,
        getDrawingLockSpy,
        () => false);
    const infiniteContext = new InfiniteContext(undefined, viewbox, cssLengthConverterFactory);
    return {
        width,
        height,
        releaseDrawingLockSpy,
        getDrawingLockSpy,
        measurementProvider,
        contextMock,
        config,
        mockDrawingIterationProvider,
        viewbox,
        infiniteContext
    };
}