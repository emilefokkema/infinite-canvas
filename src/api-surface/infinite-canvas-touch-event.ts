import { InfiniteCanvasTouchList } from "./infinite-canvas-touch-list";

export interface InfiniteCanvasTouchEvent extends TouchEvent{
    readonly targetTouches: InfiniteCanvasTouchList;
    readonly changedTouches: InfiniteCanvasTouchList;
    readonly touches: InfiniteCanvasTouchList;
}

export * from './infinite-canvas-touch-list';