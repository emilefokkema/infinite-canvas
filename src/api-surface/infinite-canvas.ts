import {Config} from "./config";
import { Units } from "./units";
import {InfiniteCanvasRenderingContext2D} from "./infinite-canvas-rendering-context-2d";
import {EventMap} from "./event-map";
import { TransformationEvent } from "./transformation-event";
import { Transformed } from './transformed';

export interface InfiniteCanvasEventHandlers extends DocumentAndElementEventHandlers, GlobalEventHandlers {
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationstart} event
     */
    ontransformationstart: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationchange} event
     */
    ontransformationchange: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationend} event
     */
    ontransformationend: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.draw} event
     */
    ondraw: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.wheelignored} event
     */
    onwheelignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.touchignored} event
     */
    ontouchignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * See [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) 
     */
    addEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    /**
     * See [`removeEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener)
     */
    removeEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | EventListenerOptions): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}

export interface InfiniteCanvas extends Config, InfiniteCanvasEventHandlers, Transformed{
    /**
     * This methods return the {@link InfiniteCanvasRenderingContext2D} belonging to this instance of {@link InfiniteCanvas}
     * 
     * @param contextType for (partial) compatibility with the other [getContext()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext)
     */
    getContext(contextType?: '2d'): InfiniteCanvasRenderingContext2D;
}

export interface InfiniteCanvasCtr {
    /**
     * Creates a new instance of an {@link InfiniteCanvas}
     * 
     * @param canvas The canvas on which to draw the {@link InfiniteCanvas}
     * @param config The {@link InfiniteCanvas}'s initial configuration
     */
    new(canvas: HTMLCanvasElement, config?: Partial<Config>): InfiniteCanvas;
    prototype: InfiniteCanvas;
    /**
     * Static property with a value of {@link Units.CANVAS}
     */
    CANVAS_UNITS: Units;
    /**
     * Static property with a value of {@link Units.CSS}
     */
    CSS_UNITS: Units;
};

declare var InfiniteCanvasConstructor: InfiniteCanvasCtr;

export * from './config'
export * from './event-map'
export * from './infinite-canvas-rendering-context-2d'
export * from './units'
export default InfiniteCanvasConstructor;
