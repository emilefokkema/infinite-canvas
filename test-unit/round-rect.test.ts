import { beforeEach, describe, expect, it} from 'vitest'
import {CanvasContextMock} from "./canvas-context-mock";
import {InfiniteCanvasRenderingContext2D} from "../src/api-surface/infinite-canvas-rendering-context-2d";
import { createInfiniteCanvasTestFixture } from './infinite-canvas-test-fixture';

describe('an infinite canvas context', () => {
	let infiniteContext: InfiniteCanvasRenderingContext2D;
	let contextMock: CanvasContextMock;

    beforeEach(() => {
		({ contextMock, infiniteContext} = createInfiniteCanvasTestFixture());
	});

    describe('that draws a round rect', () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.roundRect(10, 10, 40, 40, 5)
			infiniteContext.stroke();
		})

		it("should have given these instructions", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	})

    describe('that draws a round rect with a radius that is zero', () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.roundRect(10, 10, 40, 40, [0, 5])
			infiniteContext.stroke();
		})

		it("should have given these instructions", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	})

    describe('that draws a round rect with big radii', () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.roundRect(30, 30, 80, 30, [15, 30, 30, 15]);
			infiniteContext.stroke();
		})

		it("should have scaled the radii", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	})

    describe('that draws a round rect with big radii on invisible corners', () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.roundRect(30, 30, Infinity, 30, [15, 30, 30, 15]);
			infiniteContext.stroke();
		})

		it("should not have scaled the radii", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	})

    it.each([
		[20, 20],
		[75, 45],
		[100, 100],
		[Infinity, 20],
		[Infinity, -20],
		[-Infinity, 20],
		[-Infinity, -20],
		[20, Infinity],
		[20, -Infinity],
		[-20, Infinity],
		[-20, -Infinity],
		[Infinity, Infinity],
		[Infinity, -Infinity],
		[-Infinity, Infinity],
		[-Infinity, -Infinity]
	])('should have drawn round rect like this', (w: number, h: number) => {
		const radii = [15, {x: 60, y: 30}, 15, {x: 60, y: 30}]

		infiniteContext.beginPath();
		infiniteContext.roundRect(100, 100, w, h, radii)
		infiniteContext.stroke();

		expect(contextMock.getLog()).toMatchSnapshot();
	})


})