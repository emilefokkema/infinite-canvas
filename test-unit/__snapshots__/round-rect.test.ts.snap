// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`an infinite canvas context > should have drawn round rect like this 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(104,100)",
  "context.lineTo(104,100)",
  "context.ellipse(104,108,16,8,0,4.71238898,0,false)",
  "context.lineTo(120,116)",
  "context.save()",
  "context.transform(1,0,0,1,116,116)",
  "context.arc(0,0,4,0,1.57079633,false)",
  "context.restore()",
  "context.lineTo(116,120)",
  "context.ellipse(116,112,16,8,0,1.57079633,3.14159265,false)",
  "context.lineTo(100,104)",
  "context.save()",
  "context.transform(1,0,0,1,104,104)",
  "context.arc(0,0,4,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 2`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(115,100)",
  "context.lineTo(115,100)",
  "context.lineTo(115,100)",
  "context.ellipse(115,130,60,30,0,4.71238898,0,false)",
  "context.lineTo(175,130)",
  "context.save()",
  "context.transform(1,0,0,1,160,130)",
  "context.arc(0,0,15,0,1.57079633,false)",
  "context.restore()",
  "context.lineTo(160,145)",
  "context.ellipse(160,115,60,30,0,1.57079633,3.14159265,false)",
  "context.lineTo(100,115)",
  "context.save()",
  "context.transform(1,0,0,1,115,115)",
  "context.arc(0,0,15,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 3`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(115,100)",
  "context.lineTo(140,100)",
  "context.ellipse(140,130,60,30,0,4.71238898,0,false)",
  "context.lineTo(200,185)",
  "context.save()",
  "context.transform(1,0,0,1,185,185)",
  "context.arc(0,0,15,0,1.57079633,false)",
  "context.restore()",
  "context.lineTo(160,200)",
  "context.ellipse(160,170,60,30,0,1.57079633,3.14159265,false)",
  "context.lineTo(100,115)",
  "context.save()",
  "context.transform(1,0,0,1,115,115)",
  "context.arc(0,0,15,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 4`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(106.66666667,100)",
  "context.lineTo(201.41421356,100)",
  "context.lineTo(201.41421356,120)",
  "context.lineTo(126.66666667,120)",
  "context.ellipse(126.66666667,106.66666667,26.66666667,13.33333333,0,1.57079633,3.14159265,false)",
  "context.lineTo(100,106.66666667)",
  "context.save()",
  "context.transform(1,0,0,1,106.66666667,106.66666667)",
  "context.arc(0,0,6.66666667,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 5`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(106.66666667,100)",
  "context.lineTo(201.41421356,100)",
  "context.lineTo(201.41421356,80)",
  "context.lineTo(126.66666667,80)",
  "context.ellipse(126.66666667,93.33333333,26.66666667,13.33333333,0,4.71238898,3.14159265,true)",
  "context.lineTo(100,93.33333333)",
  "context.save()",
  "context.transform(1,0,0,1,106.66666667,93.33333333)",
  "context.arc(0,0,6.66666667,3.14159265,1.57079633,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 6`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(93.33333333,100)",
  "context.lineTo(-1.41421356,100)",
  "context.lineTo(-1.41421356,120)",
  "context.lineTo(73.33333333,120)",
  "context.ellipse(73.33333333,106.66666667,26.66666667,13.33333333,0,1.57079633,0,true)",
  "context.lineTo(100,106.66666667)",
  "context.save()",
  "context.transform(1,0,0,1,93.33333333,106.66666667)",
  "context.arc(0,0,6.66666667,0,4.71238898,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 7`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(93.33333333,100)",
  "context.lineTo(-1.41421356,100)",
  "context.lineTo(-1.41421356,80)",
  "context.lineTo(73.33333333,80)",
  "context.ellipse(73.33333333,93.33333333,26.66666667,13.33333333,0,4.71238898,0,false)",
  "context.lineTo(100,93.33333333)",
  "context.save()",
  "context.transform(1,0,0,1,93.33333333,93.33333333)",
  "context.arc(0,0,6.66666667,0,1.57079633,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 8`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(104,100)",
  "context.lineTo(104,100)",
  "context.ellipse(104,108,16,8,0,4.71238898,0,false)",
  "context.lineTo(120,201.41421356)",
  "context.lineTo(104,201.41421356)",
  "context.lineTo(100,201.41421356)",
  "context.lineTo(100,104)",
  "context.save()",
  "context.transform(1,0,0,1,104,104)",
  "context.arc(0,0,4,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 9`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(104,100)",
  "context.lineTo(104,100)",
  "context.ellipse(104,92,16,8,0,1.57079633,0,true)",
  "context.lineTo(120,-1.41421356)",
  "context.lineTo(104,-1.41421356)",
  "context.lineTo(100,-1.41421356)",
  "context.lineTo(100,96)",
  "context.save()",
  "context.transform(1,0,0,1,104,96)",
  "context.arc(0,0,4,3.14159265,1.57079633,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 10`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(96,100)",
  "context.lineTo(96,100)",
  "context.ellipse(96,108,16,8,0,4.71238898,3.14159265,true)",
  "context.lineTo(80,201.41421356)",
  "context.lineTo(96,201.41421356)",
  "context.lineTo(100,201.41421356)",
  "context.lineTo(100,104)",
  "context.save()",
  "context.transform(1,0,0,1,96,104)",
  "context.arc(0,0,4,0,4.71238898,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 11`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(96,100)",
  "context.lineTo(96,100)",
  "context.ellipse(96,92,16,8,0,1.57079633,3.14159265,false)",
  "context.lineTo(80,-1.41421356)",
  "context.lineTo(96,-1.41421356)",
  "context.lineTo(100,-1.41421356)",
  "context.lineTo(100,96)",
  "context.save()",
  "context.transform(1,0,0,1,96,96)",
  "context.arc(0,0,4,0,1.57079633,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 12`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(115,100)",
  "context.lineTo(201.41421356,100)",
  "context.lineTo(201.41421356,201.41421356)",
  "context.lineTo(115,201.41421356)",
  "context.lineTo(100,201.41421356)",
  "context.lineTo(100,115)",
  "context.save()",
  "context.transform(1,0,0,1,115,115)",
  "context.arc(0,0,15,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 13`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(115,100)",
  "context.lineTo(201.41421356,100)",
  "context.lineTo(201.41421356,-1.41421356)",
  "context.lineTo(115,-1.41421356)",
  "context.lineTo(100,-1.41421356)",
  "context.lineTo(100,85)",
  "context.save()",
  "context.transform(1,0,0,1,115,85)",
  "context.arc(0,0,15,3.14159265,1.57079633,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 14`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(85,100)",
  "context.lineTo(-1.41421356,100)",
  "context.lineTo(-1.41421356,201.41421356)",
  "context.lineTo(85,201.41421356)",
  "context.lineTo(100,201.41421356)",
  "context.lineTo(100,115)",
  "context.save()",
  "context.transform(1,0,0,1,85,115)",
  "context.arc(0,0,15,0,4.71238898,true)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > should have drawn round rect like this 15`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(85,100)",
  "context.lineTo(-1.41421356,100)",
  "context.lineTo(-1.41421356,-1.41421356)",
  "context.lineTo(85,-1.41421356)",
  "context.lineTo(100,-1.41421356)",
  "context.lineTo(100,85)",
  "context.save()",
  "context.transform(1,0,0,1,85,85)",
  "context.arc(0,0,15,0,1.57079633,false)",
  "context.restore()",
  "context.moveTo(100,100)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a round rect > should have given these instructions 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(15,10)",
  "context.lineTo(45,10)",
  "context.save()",
  "context.transform(1,0,0,1,45,15)",
  "context.arc(0,0,5,4.71238898,0,false)",
  "context.restore()",
  "context.lineTo(50,45)",
  "context.save()",
  "context.transform(1,0,0,1,45,45)",
  "context.arc(0,0,5,0,1.57079633,false)",
  "context.restore()",
  "context.lineTo(15,50)",
  "context.save()",
  "context.transform(1,0,0,1,15,45)",
  "context.arc(0,0,5,1.57079633,3.14159265,false)",
  "context.restore()",
  "context.lineTo(10,15)",
  "context.save()",
  "context.transform(1,0,0,1,15,15)",
  "context.arc(0,0,5,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(10,10)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a round rect with a radius that is zero > should have given these instructions 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(45,10)",
  "context.save()",
  "context.transform(1,0,0,1,45,15)",
  "context.arc(0,0,5,4.71238898,0,false)",
  "context.restore()",
  "context.lineTo(50,50)",
  "context.lineTo(15,50)",
  "context.save()",
  "context.transform(1,0,0,1,15,45)",
  "context.arc(0,0,5,1.57079633,3.14159265,false)",
  "context.restore()",
  "context.closePath()",
  "context.moveTo(10,10)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a round rect with big radii > should have scaled the radii 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(37.5,30)",
  "context.lineTo(95,30)",
  "context.save()",
  "context.transform(1,0,0,1,95,45)",
  "context.arc(0,0,15,4.71238898,0,false)",
  "context.restore()",
  "context.lineTo(110,45)",
  "context.save()",
  "context.transform(1,0,0,1,95,45)",
  "context.arc(0,0,15,0,1.57079633,false)",
  "context.restore()",
  "context.lineTo(37.5,60)",
  "context.save()",
  "context.transform(1,0,0,1,37.5,52.5)",
  "context.arc(0,0,7.5,1.57079633,3.14159265,false)",
  "context.restore()",
  "context.lineTo(30,37.5)",
  "context.save()",
  "context.transform(1,0,0,1,37.5,37.5)",
  "context.arc(0,0,7.5,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(30,30)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a round rect with big radii on invisible corners > should not have scaled the radii 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(45,30)",
  "context.lineTo(201.41421356,30)",
  "context.lineTo(201.41421356,60)",
  "context.lineTo(45,60)",
  "context.save()",
  "context.transform(1,0,0,1,45,45)",
  "context.arc(0,0,15,1.57079633,3.14159265,false)",
  "context.restore()",
  "context.lineTo(30,45)",
  "context.save()",
  "context.transform(1,0,0,1,45,45)",
  "context.arc(0,0,15,3.14159265,4.71238898,false)",
  "context.restore()",
  "context.moveTo(30,30)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;
